### PR TITLE
Followup #667, #714: Make it easier to switch between PHP versions

### PR DESCRIPTION
**Pull request by oxyc**
Doesn't work but I'm pushing to see if I can get travis to help out.

Sorry, I made a duplicate PR for this because rebasing caused a headache.

### CODE REVIEW

**Comment by oxyc** [comment]
Aha, the matrix didn't work as I thought it did.

**Comment by oxyc** [comment]
Tested locally:
- [x] Ubuntu 16.04: First provision using PHP 7.0
- [x] Ubuntu 16.04: First provision using PHP 7.0, rebuild using 5.6

**Comment by oxyc** [comment]
@geerlingguy do you want me to add any test for building with another version than the default? Also do you remember what the opcache.ini naming issue was?

**Comment by oxyc** [comment]
> The module purge could be avoided by setting a version specific module path so modules that require compiling like xdebug & xhprof would have a separate module path per version. The PR's referenced above would allow us to remove this purge task -- https://github.com/beetboxvm/beetbox/blob/master/provisioning/ansible/roles/beetbox-php-ppa/tasks/main.yml#L18 - @thom8 [(ref)](https://github.com/geerlingguy/drupal-vm/pull/714#issuecomment-226150941)

https://github.com/geerlingguy/ansible-role-php-xdebug/pull/32
https://github.com/geerlingguy/ansible-role-php-xhprof/pull/16

**Comment by geerlingguy** [comment]
@oxyc - Working through the linked issues :)

I'd say we should do a build for PHP 5.6 at least, maybe on Ubuntu 14.04 (or 16.04 if it's easier), since that will be an LTS we should support for a while.

**Comment by geerlingguy** [comment]
Also, as a note—I think I'd be okay with some 7.1 failures (and some failures switching versions here and there), and we can call the 'version switching' an experimental feature for now. Best option is to always destroy and rebuild (like we always have done), but you can try switching PHP versions on the fly if you want... something to that effect.

Looks like you've already updated docs, and everything else I'm seeing looks wonderful. Plus, even if not switching on-the-fly, having the Version be a variable (by using Ondrej's repo in all cases) is quite a slick feature!

Thanks for the work on this! After I kick the tires a tiny bit more, I'm planning on merging tonight.

 🚂 All aboard the release train!

**Comment by oxyc** [comment]
Let me run a few more tests as well! I'll try and get 12.04 with redis as well as php 7.1 with xdebug working tonight.

**Comment by geerlingguy** [comment]
@oxyc - I'm all for it—you're on a roll, I don't think I want to slow you down :)

Just give me the 👍 when you're ready for me to merge this (I'm running a couple builds now).

**Comment by oxyc** [comment]
2.4mbit connection is slooooooooooow....

**Comment by oxyc** [comment]
@geerlingguy https://github.com/geerlingguy/ansible-role-php-redis/pull/2 fixed Ubuntu 12.04 for me. It runs with PHP 7.1 and redis. Let's just wait for travis to get back and then this gets my 👍

**Comment by oxyc** [comment]
That's just awesome! Thanks @thom8 for all your work on this.

**Comment by geerlingguy** [comment]
php-redis role is happy and now part of Drupal VM. Running through code one last time to make sure there's nothing major we're missing...

**Comment by geerlingguy** [comment]
Thanks to both @thom8 _and_ @oxyc — I don't think I would've ever taken the time to get this working so nicely...

You guys deserve a lot of 🍻 !

File: docs/other/php.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+Drupal VM defaults to PHP 7.0, but you can also install and use 5.6 or 7.1.
+
+## Ubuntu
+
+Ondřej Surý's PPA for PHP is used to install PHP 7.0, but you can switch to using 5.6 or 7.1 packages by changing `php_version` inside `config.yml` to eg. `"7.1"`.
+
+If you're using Apache with `mod_php` you should also add `libapache2-mod-php7.1` to the `extra_packages` list.
```
Comment (geerlingguy) [inline docs/other/php.md]: We need to specify what to do for PHP 5.6 too—or reword this so it is less 7.1-specific.

File: examples/acquia/acquia.config.yml [hunk 2]
```diff
@@ -4,38 +4,6 @@ vagrant_box: geerlingguy/ubuntu1204
 
 # Acquia Cloud defaults to PHP 5.6.
 php_version: "5.6"
-php_install_recommends: no
```
Comment (geerlingguy) [inline examples/acquia/acquia.config.yml]: Reminder to self—I'll need to update BLT's internal config.yml once this stuff lands.

**Comment by geerlingguy** [comment]
MERGED!

**Comment by geerlingguy** [comment]
Now, to retest things a couple more times... I'll either tag 4.0.0 tomorrow sometime (I have a car inspection scheduled...), or on Monday when I'm prepping for a trip to Baltimore.

This one's getting a Twitter mention, a Blog post, the whole nine yards!

**Comment by sbubaron** [comment]
FYI After switching to 5.6 from 7 I was getting the error:
``` 
Failed loading /usr/lib/php5/modules/xdebug.so:  /usr/lib/php5/modules/xdebug.so: undefined symbol: zend_sort
```
A vagrant destroy and a vagrant up fixed it. Might want to add that to the docs and maybe change the release notes so that it doesn't imply you can just provision. 

Also the URL in default.config.vm docs is broken: 
To use 5.6, see: http://docs.drupalvm.com/en/latest/other/php-56/ should be http://docs.drupalvm.com/en/latest/other/php/

Finally, I had previously specified the following php_packages in my config.yml (I don't remember why, I imagine I copied them from default.config.yml at somepoint) -- are these no longer needed in Drupal-VM 4.0? Or should I update them to reference the php5.6 versions?

```
php_packages:
  - php7.0-common
  - php7.0-cli
  - php7.0-dev
  - php7.0-fpm
  - libpcre3-dev
  - php7.0-gd
  - php7.0-curl
  - php7.0-imap
  - php7.0-json
  - php7.0-opcache
  - php7.0-xml
  - php7.0-mbstring
  - php-sqlite3
  - php-apcu
  - php7.0-mcrypt
```

**Comment by oxyc** [comment]
> Failed loading /usr/lib/php5/modules/xdebug.so:  /usr/lib/php5/modules/xdebug.so: undefined symbol: zend_sort

I [had this at some point](https://github.com/geerlingguy/drupal-vm/pull/1043#issuecomment-266140659) but couldn't reproduce it so I thought it was fixed somehow. Might be related and fixable now that geerlingguy/ansible-role-php-xdebug#32 geerlingguy/ansible-role-php-xhprof#16 are merged. Would you mind creating a separate issue?

> Also the URL in default.config.vm docs is broken:
To use 5.6, see: http://docs.drupalvm.com/en/latest/other/php-56/ should be http://docs.drupalvm.com/en/latest/other/php/

This is already fixed in master. https://github.com/geerlingguy/drupal-vm/commit/c626a3e5cb0800f2ba163bec9d580597d98fa210

> Finally, I had previously specified the following php_packages in my config.yml (I don't remember why, I imagine I copied them from default.config.yml at somepoint) -- are these no longer needed in Drupal-VM 4.0? Or should I update them to reference the php5.6 versions?

- `libpcre3-dev` is a [dependency of php*-dev](https://packages.debian.org/sid/php5-dev) so doesn't need to be specified. I excluded it so it's easier to purge the packages.
- `php-sqlite3`, `php-apcu` are found using `php7.0-*` as well.
- The rest installed.

**Pull request merged**